### ISSUE TITLE
cuDNN, compile errors

### ISSUE THREAD

**Issue by mynameis2** [dmlc/cxxnet#42]
compile with cuDNN library provide error below..  without cudnn all compiles ok, 660M nvidia card(compute capability 3.0). Using cuDNN v2 Release Candidate 3. With cuDNN 6.5 R1 - similar errors... any helpful ideas?

nvcc -c -o layer_gpu.o --use_fast_math -g -O3 -ccbin g++  -Xcompiler "-DMSHADOW_FORCE_STREAM -Wall -g -O3 -I./mshadow/  -fopenmp -fPIC -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_DIST_PS=0 -DCXXNET_USE_OPENCV=1 -DCXXNET_USE_CUDNN=1 -I/usr/local/cuda-7.0/include/ -L/usr/local/cuda-7.0/lib64/ " src/layer/layer_impl.cu
src/layer/./cudnn_pooling_layer-inl.hpp(84): error: identifier "CUDNN_POOLING_AVERAGE" is undefined
          detected during:
            instantiation of "void cxxnet::layer::CuDNNPoolingLayer<Reducer, mode, xpu>::InitConnection(const std::vector<cxxnet::layer::Node<xpu> *, std::allocator<cxxnet::layer::Node<xpu> *>> &, const std::vector<cxxnet::layer::Node<xpu> *, std::allocator<cxxnet::layer::Node<xpu> *>> &, cxxnet::layer::ConnectState<xpu> *) [with Reducer=mshadow::red::maximum, mode=11, xpu=cxxnet::gpu]" 
(16): here
            instantiation of "cxxnet::layer::CuDNNPoolingLayer<Reducer, mode, xpu>::CuDNNPoolingLayer() [with Reducer=mshadow::red::maximum, mode=11, xpu=cxxnet::gpu]" 
src/layer/layer_impl-inl.hpp(55): here
            instantiation of "cxxnet::layer::ILayer<xpu> *cxxnet::layer::CreateLayer_(cxxnet::layer::LayerType, mshadow::Random<xpu, mshadow::default_real_t> *, const cxxnet::layer::LabelInfo *) [with xpu=cxxnet::gpu]" 
src/layer/layer_impl.cu(12): here

1 error detected in the compilation of "/tmp/tmpxft_00003eba_00000000-7_layer_impl.cpp1.ii".
make: *** [layer_gpu.o] Error 2

**Comment by antinucleon** [comment]
I am using CuDNN R2 RC2, I didn't track RC3 yet since NVIDIA change its interface frequently. I will track it sometime today. From error information, seems NVIDIA removes CUDNN_POOLING_AVERAGE. You can try to remove AvgPooling by delete this line: https://github.com/antinucleon/cxxnet/blob/V2-refactor/src/layer/layer_impl-inl.hpp#L57

**Comment by mynameis2** [comment]
Thank you for answer! Removing lines had no effect, error stay the same..   I found similar problems: https://github.com/BVLC/caffe/pull/1739 & https://github.com/Theano/Theano/issues/2481  
Very strange, but on nvidia site i can not find CuDNN R2 RC2, only cuDNN v1 (cuDNN 6.5 R1) and cuDNN v2 Release Candidate 3 (Feb 26,2015). If it possible, could you provide a link to cudnn-6.5-linux-x64-v2-rc2.tgz if you keep it file?

**Comment by antinucleon** [comment]
You can simply change CUDNN_POOLING_AVERAGE to CUDNN_POOLING_AVERAGE_COUNT_EXCLUDE_PADDING at https://github.com/antinucleon/cxxnet/blob/V2-refactor/src/layer/cudnn_pooling_layer-inl.hpp#L84  to make it compile

**Comment by mynameis2** [comment]
well, i compile cxxnet without errors with cudnn, and training become ~25% faster) but for some reason training don't going well(without cudnn everything fine)... 
round        0:[     460] 105 sec elapsed[1]    train-error:0.998743	val-error:0.999558
round        1:[     460] 241 sec elapsed[2]    train-error:0.999558	val-error:0.999558
But still thank you for replying! cxxnet is a great work!

**Comment by antinucleon** [comment]
Could you share me your configuration file? I didn't test RC3 yet since I don't have empty machine now. I will test RC3 in this evening.

**Comment by mynameis2** [comment]
sorry, I was asleep.

train.conf:
data = train
iter = imgbin
  image_list = "./tr.lst"
  image_bin  = "./tr.bin"
  image_mean = "models/image_mean.bin"
  rand_mirror=1                #commenting this not help
  rand_crop=1                  #commenting this not help
  max_rotate_angle=180 #commenting this not help
  max_aspect_ratio = 0.5 #commenting this not help
  max_shear_ratio = 0.3   #commenting this not help
  min_crop_size=32          #commenting this not help
  max_crop_size=40         #commenting this not help
iter = threadbuffer
iter = end

eval = val
iter = imgbin
  image_list = "./tr.lst"
  image_bin  = "./tr.bin"
  image_mean = "models/image_mean.bin"
iter = threadbuffer
iter = end

netconfig=start
layer[+1] = conv
  kernel_size = 4
  stride = 1
  nchannel = 48
  pad = 2
layer[+1] = relu
layer[+1] = max_pooling
  kernel_size = 3
  stride = 2
###############
layer[+1] = conv
  nchannel = 96
  kernel_size = 3
  stride = 1
  pad = 1
layer[+1] = relu
layer[+1] = conv
  nchannel = 96
  kernel_size = 3
  stride = 1
  pad = 1
layer[+1] = relu
layer[+1] = max_pooling
  kernel_size = 3
  stride = 2
##############
layer[+1] = conv
  nchannel = 128
  kernel_size = 2
  stride = 1
layer[+1] = relu
layer[+1] = conv
  nchannel = 128
  kernel_size = 3
  stride = 1
layer[+1] = max_pooling
  kernel_size = 3
  stride = 2
layer[+1] = flatten
layer[+1] = fullc
  nhidden = 256
layer[+0] = dropout
  threshold = 0.5
layer[+1] = fullc
  nhidden = 121
layer[+0] = softmax
netconfig=end

# evaluation metric
metric = error

max_round = 1000
num_round = 1000

# input shape not including batch
input_shape = 3,40,40

batch_size = 64

# global parameters in any sectiion outside netconfig, and iter
momentum = 0.9
wmat:lr  = 0.001
wmat:wd  = 0.0005

bias:wd  = 0.000
bias:lr  = 0.002

# all the learning rate schedule starts with lr
lr:schedule = expdecay
lr:gamma = 0.1
lr:step = 20000

save_model=1
model_dir=models
print_step = 1
# random config
random_type = xavier
init_sigma = 0.01

# new line


and in config.mk:
USE_BLAS = openblas
USE_CUDNN = 1
ADD_CFLAGS = -I/usr/local/cuda-7.0/include/ -L/usr/local/cuda-7.0/lib64/

**Comment by antinucleon** [comment]
The problem may related to CuDNN max pooling. Please try newest version (cudnn maxpooling disabled)
